test/rematch2.jl

```diff
@@ -1,13 +1,71 @@
-# Note we do not use `@eval` to define a struct within a @testset
+# Note we do not use `@eval` to define types within a `@testset``
 # because we need the types to be defined during macro expansion,
 # which is earlier than evaluation.  types are looked up during
-# expansion of the @match2 macro so we can use the known bindings
+# expansion of the `@match2`` macro so we can use the known bindings
 # of types to generate more efficient code.
 
+file = Symbol(@__FILE__)
+
 @testset "@rematch2 tests" begin
 
-    @testset "Assignments in the value DO leak out (when not using `let``)" begin
-        @match2 Foo(1, 2) begin
+@testset "Check that `where` clauses are reparsed properly 1" begin
+    x = true
+    @test (@match2 3 begin
+        ::Int where x => 1
+        _ => 2
+    end) == 1
+
+    x = false
+    @test (@match2 3 begin
+        ::Int where x => 1
+        _ => 2
+    end) == 2
+end
+
+@testset "Check that `where` clauses are reparsed properly 2" begin
+    x = true
```

Comment:
could you add a test along the lines of
```
x = 2
@test (@match2 3 begin
   a::Int where x == a => a
   _ => 2
end)
```
and the `!=` case as well? (or any falsey case)

Reply:
I have a sinking feeling it won't work right.  Maybe we should move back to the `pattern if condition end` syntax from `Match.jl`.  That syntax is unambiguous.

Reply:
Huh.  It works.